linux-gpib-user/doc/man/man3/ibdev.3
'\" t
.\"     Title: ibdev
.\"    Author: Frank Mori Hess
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 10/04/2025
.\"    Manual: 	Traditional API Functions 
.\"    Source: linux-gpib 4.3.7
.\"  Language: English
.\"
.TH "IBDEV" "3" "10/04/2025" "linux-gpib 4.3.7" "Traditional API Functions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ibdev \- open a device (device)
.SH "SYNOPSIS"
.sp
.ft B
.nf
#include <gpib/ib\&.h>
.fi
.ft
.HP \w'int\ ibdev('u
.BI "int ibdev(int\ " "board_index" ", int\ " "pad" ", int\ " "sad" ", int\ " "timeout" ", int\ " "send_eoi" ", int\ " "eos" ");"
.SH "DESCRIPTION"
.PP
ibdev() is used to obtain a device descriptor, which can then be used by other functions in the library\&. The argument
\fIboard_index\fR
specifies which GPIB interface board the device is connected to\&. The
\fIpad\fR
and
\fIsad\fR
arguments specify the GPIB address of the device to be opened (see
ibpad()
and
ibsad())\&. The timeout for io operations is specified by
\fI timeout \fR
(see
ibtmo())\&. If
\fIsend_eoi\fR
is nonzero, then the EOI line will be asserted with the last byte sent during writes (see
ibeot())\&. Finally, the
\fIeos\fR
argument specifies the end\-of\-string character and whether or not its reception should terminate reads (see
ibeos())\&. To suppress error messages printed to stderr by ibdev() set the environment variable IB_NO_ERROR before calling the routine\&.
.SH "RETURN VALUE"
.PP
If sucessful, returns a (non\-negative) device descriptor\&. On failure, \-1 is returned\&.
.SH "COPYRIGHT"
.br
Copyright \(co 2003-2006, 2008 Frank Mori Hess
.br


ChangeLog
2018-12-25 09:50  dpenkler

	* [r1806] Fix board type

2018-12-25 09:46  dpenkler

	* [r1805] Fix ni_usb_gpib cut & paste error

2018-12-25 09:42  dpenkler

	* [r1804] Simplify usb udev rules as per Frank's suggestion

2018-12-21 10:44  dpenkler

	* [r1803] Changes for multiple board configuration

	  Update INSTALL
	  Add serial to device environment
	  Print serial in gpib_udev_config
	  Add example rules for agilent

2018-12-15 08:55  fmhess

	* [r1802] Add some examples for configuring multiple boards with
	  different minor numbers.

2018-12-15 08:29  fmhess

	* [r1801] Automatically gibe the correct board type to gpib_config
	  in udev rules

2018-12-15 08:01  fmhess

	* [r1800] rename gpib generic rules so it runs at 98
	  (before the driver-specific fules which run at 99)

2018-12-15 07:52  fmhess

	* [r1799] Add fix for mylog from gpib_udev_config to
	  gpib_udev_fxloader.in. Also add some quoting to
	  ensure whitespace is preserved in log messages.

2018-12-14 16:23  dpenkler

	* [r1798] Fix logging in gpib_udev_config for gpib_config options
	  --

2018-12-14 08:30  fmhess

	* [r1797] Removed some bash-isms so scripts work with posix
	  /bin/sh.

2018-12-14 08:11  fmhess

	* [r1796] Revert commit 1793

2018-12-14 08:07  fmhess

	* [r1795] Add --with-udev-libdir option to configure, that I failed
	  to include in earlier commit.

2018-12-13 17:05  dpenkler

	* [r1794] Really remove unused files

2018-12-13 17:02  dpenkler

	* [r1793] Add UDEV_LIBDIR to Makefile.ams\nRemove unused files

2018-12-13 13:44  fmhess

	* [r1792] Added --with-udev-libdir configure option, defaulted
	  to PREFIX/lib/udev.

2018-12-13 12:24  dpenkler

	* [r1791] Rename and move udevadm wrapper to libdir

2018-12-13 12:17  fmhess

	* [r1790] Add missing define of UDEV_LIB_DIR to prevent files from
	  getting installed into root directory.

2018-12-13 12:13  fmhess

	* [r1789] Replaced udev scripts ni_usb_gpib and agilent_82357a with
	  a single generic script.

2018-12-13 11:53  fmhess

	* [r1788] Really delete usermap files this time.

2018-12-13 11:47  fmhess

	* [r1787] Pull out firmware loading into a single udev script
	  gpib_udev_fxloader.

2018-12-13 11:44  dpenkler

	* [r1786] Use #!/bin/sh instead of #!/bin/bash

2018-12-13 11:11  dpenkler

	* [r1785] Add wrapper script for udevadm

2018-12-12 18:04  dpenkler

	* [r1784] Fix udev rules scripts for agilent usb

2018-12-12 15:43  fmhess

	* [r1783] Generate usb script/rules during "make all" rather
	  than possibly waiting until "make install", where the
	  user may be running as root.

2018-12-12 15:27  fmhess

	* [r1782] Fixed assignment that was supposed to be a comparison

2018-12-12 14:03  fmhess

	* [r1781] Removed obsolete hotplug usermaps.

	  Fixed problems with usb udev rules not working if the driver
	  module was not already loaded when the hardware was plugged
	  in.

	  Made usb drivers respond to the sysfs device path of their
	  usb interface rather than the parent usb device. This is helpful
	  since the driver binds to the interfaces rather than the device.

2018-12-12 06:25  fmhess

	* [r1780] Update comments in agilent_82357a udev script to use
	  new --serial and --sysfs-device-path gpib_config options.

2018-12-10 10:49  fmhess

	* [r1779] Added linux/atomic.h compatibility header.

2018-12-10 10:36  fmhess

	* [r1778] Added a couple more devices to ni_usb_gpib.usermap.

2018-12-10 10:28  fmhess

	* [r1777] Add kernel driver support for selecting hardware by sysfs
	  device path
	  or serial number.

2018-12-10 10:27  fmhess

	* [r1776] Add support for selecting hardware by sysfs device path
	  or serial number. Updated gpib_config docs with
	  missing info on some options. Noted in docs that
	  gpib.conf will be in the sysconfdir, which isn't
	  always /etc/ depending on configuration.

2018-12-10 10:11  fmhess

	* [r1775] Add device ids for various ni adapters to
	  ni_usb_gpib.usermap

2018-12-07 09:12  dpenkler

	* [r1774] Remove dependency on autoconf tools
	  Modify INSTALL instructions
	  Update drivers/gpib/Makefile
	  Add Makefile to invoke kernel build scripts
	  Add version.h to drivers/gpib/include
	  Make definitions of smp_mb__xxx_atomic independent of kernel
	  version
	  Add VERBOSE flag for make, default is VERBOSE=0
	  so that build log is more overseeable.
	  Add HAVE_DEV_OF_NODE workaround for CentOS/RHEL7
	  Fix compile warnings in lpvo

2018-12-06 06:10  fmhess

	* [r1773] Reverted r1766 since it was actually comparing the serial
	  number to the device tree path. Linux doesn't seem to
	  have a universal (for all bus types) way of specifying
	  a piece of hardware by path, so we can't use the
	  device tree path to specify a usb device anyways.

2018-12-05 11:19  dpenkler

	* [r1772] Apply gpib_lines patch from Tomislav Ivek

2018-11-29 08:29  dpenkler

	* [r1771] Change barrier() to mb() to be generic

2018-11-28 19:02  dpenkler

	* [r1770] Add kernel version ifdef for smp atomics

2018-11-28 18:02  dpenkler

	* [r1769] Add compat fixes for kernels < 3.15 and 3.16

2018-11-28 14:36  dpenkler

	* [r1768] Add patch from Tomislav Ivek

	  The current linux-gpib Python class can run into a double-close
	  situation. This bug is related to a design point in Gpib.Gpib
	  which
	  cleans up its GPIB handle only on Python's GC cycle. This is not
	  guaranteed to run when the programmer expects it to and in
	  extreme
	  cases might even leak resources. This is resolved by adding a
	  Gpib.close() method for user-controlled cleanup.

2018-11-28 10:26  fmhess

	* [r1767] Applied patch from Derek Kozel <derek.kozel@gmail.com>:

	  During the linux-gpib-kernel make install process the kernel
	  modules_install command is run bit the DESTDIR argument is not
	  passed
	  in meaning that it will always try to install to the system
	  directory
	  rather than a custom directory. Adding
	  INSTALL_MOD_PATH="$(DESTDIR)" to
	  the command correctly passes the information.

2018-11-16 20:30  fmhess


linux-gpib-user/lib/ibdev.c
/***************************************************************************
                                 ibdev.c
                             -------------------
    begin                : Tues Feb 12 2002
    copyright            : (C) 2002 by Frank Mori Hess
    email                : fmhess@users.sourceforge.net
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "ib_internal.h"
#include <stdlib.h>

#if 0
static int is_device_addr(int minor, int pad, int sad)
{
	ibBoard_t *board;
	unsigned int board_pad;
	int board_sad;

	board = &ibBoard[ minor ];

	if (query_pad(board, &board_pad) < 0) {
		fprintf(stderr, "failed to query pad\n");
		return -1;
	}
	if (query_sad(board, &board_sad) < 0) {
		fprintf(stderr, "failed to query sad\n");
		return -1;
	}

	if (gpib_address_equal(board_pad, board_sad, pad, sad) == 0) {
		return 1;
	}

	return 0;
}
#endif

int ibdev(int minor, int pad, int sad, int timo, int eot, int eosmode)
{
	int retval,ud;
	ibConf_t new_conf;
	int no_show_error = getenv("IB_NO_ERROR") ? 1 : 0;
	ibBoard_t *board;
	unsigned  board_pad;
	int my_sad, board_sad;

	if (minor < 0 || minor >=  GPIB_MAX_NUM_BOARDS) {
		if (!no_show_error)
			fprintf(stderr,"ibdev: invalid minor%d\n", minor);
		setIberr(EARG);
		sync_globals();
		return -1;
	}

	retval = ibParseConfigFile(minor);
	if (retval < 0) {
		if (errno) {
			setIberr(EDVR);
			setIbcnt(errno);
		} else {
			setIberr(ECNF);
		}
		setIbsta(ERR);
		sync_globals();
		return -1;
	}

	/*  Check for valid address arguments */

	if (pad < 0 || pad > gpib_addr_max) {
		if (!no_show_error)
			fprintf(stderr,"ibdev: invalid pad %d, expected 0 <= pad <= 30\n", pad);
		setIberr(EARG);
		sync_globals();
		return -1;
	}

	if (!sad) /* 0 means no address */
		my_sad = -1;
	else
		my_sad = sad - sad_offset;

	if (my_sad < -1 || my_sad > gpib_sad_max) {
		if (!no_show_error)
			fprintf(stderr,"ibdev: invalid sad 0x%02x, expected 0, or 0x60 <= sad <= 0x7f\n", sad);
		setIberr(EARG);
		sync_globals();
		return -1;
	}

	init_ibconf(&new_conf);
	new_conf.settings.pad = pad;
	new_conf.settings.sad = my_sad;                     /* device address                   */
	new_conf.settings.board = minor;                    /* board number                     */
	new_conf.settings.eos = eosmode & 0xff;             /* local eos modes                  */
	new_conf.settings.eos_flags = eosmode & 0xff00;
	new_conf.settings.usec_timeout = timeout_to_usec(timo);
	if (eot)
		new_conf.settings.send_eoi = 1;
	else
		new_conf.settings.send_eoi = 0;
	new_conf.defaults = new_conf.settings;
	new_conf.is_interface = 0;
	new_conf.error_msg_disable = no_show_error;

	ud = my_ibdev(new_conf);

	if (ud < 0)
		return -1;

// check for address conflicts with board addresses

	board = interfaceBoard(&new_conf);

	if (query_pad(board, &board_pad) < 0) {
		goto ibdev_err;
	}
	if (query_sad(board, &board_sad) < 0) {
		goto ibdev_err;
	}

	if (board_pad == pad) {
		fprintf(stderr,"ibdev: address conflict with board pad\n");
		setIberr(EARG);
		goto ibdev_err;
	}

	return ud;
ibdev_err:
	retval = close_gpib_handle(&new_conf);
	if (retval < 0) {
		if (!no_show_error)
			fprintf(stderr, "ibdev: cleanup failed\n");
		sync_globals();
		return -1;
	}
	release_descriptor(ud);
	sync_globals();
	return -1;
}

int my_ibdev(ibConf_t new_conf)
{
	int ud;
	ibConf_t *conf;

	ud = insert_descriptor(new_conf, -1);
	if (ud < 0) {
		if (!new_conf.error_msg_disable)
			fprintf(stderr, "libgpib: ibdev failed to get descriptor\n");
		setIbsta(ERR);
		return -1;
	}

	conf = enter_library(ud);
	if (!conf) {
		exit_library(ud, 1);
		release_descriptor(ud);
		return -1;
	}
	// XXX do local lockout if appropriate

	exit_library(ud, 0);
	return ud;
}
